url_matcher/matcher.py

```diff
@@ -9,11 +9,28 @@
 from url_matcher.util import get_domain
 
 
-@dataclass
+@dataclass(init=False, frozen=True)
 class Patterns:
-    include: List[str]
-    exclude: List[str] = field(default_factory=list)
-    priority: int = 500
+    include: Tuple[str, ...]
+    exclude: Tuple[str, ...]
+    priority: int
+
+    def __init__(self, include: List[str], exclude: Optional[List[str]] = None, priority: int = 500):
+        # The initialization is manually set so that we can support an API of
+        # accepting and returning lists. However, tuples are being used underneath
+        # that class so that the attributes are truly immutable, in addition to
+        # being frozen=True.
+        # Using lists are far less likely to have human typing mistakes compared to
+        # tuples since the trailing `,` char can easily be missed out. For
+        # example:
+        #     *  ("element") is not the same as ("element",) which is a tuple.
+        # Lastly, the manner of how we set the attribute values below is in line
+        # with how Python's own `dataclasses` library assign attributes to frozen
+        # classes. Here's a reference:
+        #     * https://github.com/python/cpython/blob/main/Lib/dataclasses.py#L1119-L1122
```

Comment:
could you please use a link which includes a commit hash?

Reply:
Good call! Used a version tag of `3.10.2` instead to better clarify the intended Python version. https://github.com/zytedata/url-matcher/pull/4/commits/55a61e095e4acc123c50531cb593722d01d40027